crates/autopilot/src/decoded_settlement.rs

```diff
@@ -280,91 +268,49 @@ impl DecodedSettlement {
-    }
-
-    /// Returns the list of executions with their fees,
-    /// which are supposed to be updated whenever a new settlement is executed.
-    /// Done for the orders that have solver-computed fees (limit orders).
-    pub fn order_executions(
+    /// Returns unsubsidized fees for all trades.
+    /// Length of the returned vector is equal to the length of `self.trades`
+    /// and `order_fees`.
+    pub fn all_fees(
         &self,
         external_prices: &ExternalPrices,
-        mut orders: Vec<OrderExecution>,
+        order_fees: &[(OrderUid, Option<U256>)],
     ) -> Vec<Fees> {
         self.trades
             .iter()
-            .filter_map(|trade| {
-                let i = orders.iter().position(|order| {
-                    trade.matches_execution(order, &self.tokens, &self.domain_separator)
-                })?;
-                // It's possible to have multiple fills with the same `executed_amount` for
-                // the same order with different `solver_fees`. To
-                // end up with the correct total fees we can only
-                // use every `OrderExecution` exactly once.
-                let order = orders.swap_remove(i);
-
-                // Update fee only for orders where fee is taken from surplus
-                if !matches!(order.executed_fee, ExecutedFee::Surplus) {
-                    return None;
-                }
-
-                let fees = self.fee(external_prices, &order, trade);
-
-                if fees.is_none() {
-                    tracing::warn!("possible incomplete fee calculation");
+            .zip(order_fees.iter())
+            .map(|(trade, (order, order_fee))| {
+                match self.fee(external_prices, *order, *order_fee, trade) {
+                    Some(fees) => fees,
+                    None => {
+                        tracing::warn!("possible incomplete fee calculation");
+                        // we should have an order execution for every trade
+                        Fees {
+                            order: *order,
+                            sell: U256::zero(),
+                            native: U256::zero(),
+                        }
+                    }
                 }
-                fees
             })
             .collect()
     }
 
     fn fee(
         &self,
         external_prices: &ExternalPrices,
-        order: &OrderExecution,
+        order: OrderUid,
+        order_fee: Option<U256>,
         trade: &DecodedTrade,
     ) -> Option<Fees> {
         let sell_index = trade.sell_token_index.as_u64() as usize;
         let buy_index = trade.buy_token_index.as_u64() as usize;
         let sell_token = self.tokens.get(sell_index)?;
         let buy_token = self.tokens.get(buy_index)?;
 
-        let solver_fee = match &order.executed_fee {
-            ExecutedFee::Order(fee) => *fee,
-            ExecutedFee::Surplus => {
+        let fee = match order_fee {
+            Some(fee) => fee,
+            None => {
```

Comment:
Would it make sense to add the order_fee and surplus_fee in case both fields are populated? Would this be more or less error prone?

Reply:
This is a bit more involving change than just adding the two. The `fee` function is used both for total fees and order executions (which should contain only surplus fees). I do have in plan to support limit orders with `fee_amount` !=0 but still need to finish the fee simplification (it's not urgent since we still make sure that type of orders are not even created).